torch/nn/modules/linear.py

```diff
@@ -115,4 +116,99 @@ def __repr__(self):
+        >>> output = m(input)
+        >>> print(output.size())
+    """
+    def __init__(self, in_features, out_features, bias=True, factorised=True, std_init=None):
+        super(NoisyLinear, self).__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+        self.factorised = factorised
+        self.weight_mu = Parameter(torch.Tensor(out_features, in_features))
+        self.weight_sigma = Parameter(torch.Tensor(out_features, in_features))
+        if bias:
+            self.bias_mu = Parameter(torch.Tensor(out_features))
+            self.bias_sigma = Parameter(torch.Tensor(out_features))
+        else:
+            self.register_parameter('bias', None)
+        if not std_init:
+            if self.factorised:
+                self.std_init = 0.4
+            else:
+                self.std_init = 0.017
+        else:
+            self.std_init = std_init
+        self.reset_parameters(bias)
+        self.reset_noise()
+
+    def reset_parameters(self, bias):
+        if self.factorised:
+            mu_range = 1. / math.sqrt(self.weight_mu.size(1))
+            self.weight_mu.data.uniform_(-mu_range, mu_range)
+            self.weight_sigma.data.fill_(self.std_init / math.sqrt(self.weight_sigma.size(1)))
+            if bias:
+                self.bias_mu.data.uniform_(-mu_range, mu_range)
+                self.bias_sigma.data.fill_(self.std_init / math.sqrt(self.bias_sigma.size(0)))
+        else:
+            mu_range = math.sqrt(3. / self.weight_mu.size(1))
+            self.weight_mu.data.uniform_(-mu_range, mu_range)
+            self.weight_sigma.data.fill_(self.std_init)
+            if bias:
+                self.bias_mu.data.uniform_(-mu_range, mu_range)
+                self.bias_sigma.data.fill_(self.std_init)
+
+    def scale_noise(self, size):
+        x = torch.Tensor(size).normal_()
+        x = x.sign().mul(x.abs().sqrt())
+        return x
+
+    def reset_noise(self):
+        if self.factorised:
+            epsilon_in = self.scale_noise(self.in_features)
+            epsilon_out = self.scale_noise(self.out_features)
+            self.weight_epsilon = Variable(epsilon_out.ger(epsilon_in))
+            self.bias_epsilon = Variable(self.scale_noise(self.out_features))
+        else:
+            self.weight_epsilon = Variable(torch.Tensor((self.out_features, self.in_features)).normal_())
+            self.bias_epsilon = Variable(torch.Tensor(self.out_features).normal_())
+
+    def forward(self, input):
+        if self.training:
+            return F.linear(input,
+                            self.weight_mu + self.weight_sigma.mul(self.weight_epsilon),
+                            self.bias_mu + self.bias_sigma.mul(self.bias_epsilon))
+        else:
+            return F.linear(input, self.weight_mu, self.bias_mu)
+
+    def __repr__(self):
+        return self.__class__.__name__ + ' (' \
+            + str(self.in_features) + ' -> ' \
+            + str(self.out_features) + ')'
```

Comment:
It's probably worth including `self.factorised` in this as well. `self.bias` is left out of `nn.Linear`, and `self.std_init` is only applicable at init.